### DISCUSSION TITLE
Fortran HDF5 output writing

### DISCUSSION THREAD

**Discussion by LordOfBunnies** [AMReX-Codes/amrex#2870] [Q&A]
I'm currently writing an entropic LBM flow solver using Amrex to do the AMR work.  I was originally using CGNS to do the output, but weird malloc errors kept that from working once the AMR was switched on.  

I'm currently trying to get Amrex to do HDF5 output as I haven't been able to get anything capable of doing normal Amrex post-processing to work on my workstation.  I've been using Tecplot to do visualization from CGNS (when that was working), and Tecplot will handle HDF5 output.  That being said, I modified the code a little to allow my Fortran code to use the Amrex's HDF5 output routines.  Amrex is version 22.06, HDF5 is 1.12.2, OpenMPI is 4.0.1, GCC is 8.3.1 and I compiled everything with -std=c17 and -std=c++17 where relevant.  HDF5 isn't using any compression (yet, maybe later).

As it stands, the code will output a .h5 file, but attempting to read it into Tecplot produces this error:
Dimension of selected SDS are not equivalent
/level_0/data:datatype=0 IMax =131072 JMax = 1 Kmax = 1 rank = 1

Edit: A little reading says that how Amrex does its HDF5 output.  I managed to find an old version of Paraview (3.98) which will read the output, but not the time information making actually post-processing anything ridiculously difficult.

I did check and made sure that Amrex had AMREX_SPACEDIM = 3 for this case.  131072 is also the total number of cells involved on lvl 0 (128x32x32), but it seems to not be writing the correct dimension to the file since rank = 1.  The file size is also not increasing when a second iteration calls it as well.  Even with a single iteration of output, it fails the same way.

Sadly, the code is WAY too large to reproduce a minimal working example.

Do you foresee any reason HDF5 wouldn't be getting good data from Amrex?  How best can I check that everything is working correctly within Amrex's HDF5 plotfile routines?

**Comment by WeiqunZhang** [comment]
I assume you are calling from Fortran a C++ wrapper function with extern C interface that calls amrex::WriteMultiLevelPlotfileHDF5 or amrex::WriteSingleLevelPlotfileHDF5.  Maybe you can print out some information in the wrapper function.  If you like, maybe you could show us that part of the code.

**Reply by LordOfBunnies** [reply]
I'm about 90% sure the Amrex side is working correctly, but Tecplot doesn't have the required Chombo interface to read the files correctly.  If you know of any scripts, macros, etc. to convert to different HDF5 formats, that might let Tecplot work with it.  Unfortunately, I'm on a workstation on campus, so I need to pester our IT guy to see about updating OpenGL to get the more recent versions of Paraview to function, or updating the compiler so GLibC gets to a recent enough version for VISiT to work, or successfully install the open-motif dev pack so something else will compile.

After I realized what was going wrong, I wanted to delete the topic, but I haven't figure out how yet.  It's possible I'm missing a compile flag in HDF5 or Amrex to make it work, but I'm less confident in that possibility.

As for the files, I've attached them here.  If you want to add the options via proper Github forking methods, that might take me a hot second to figure out.  The current plan is serialize the output and just do Plot3D files, because I know those work with things I have access to.
[AMReX_plotfile_fi.txt](https://github.com/AMReX-Codes/amrex/files/9060267/AMReX_plotfile_fi.txt)
[AMReX_plotfile_mod.txt](https://github.com/AMReX-Codes/amrex/files/9060268/AMReX_plotfile_mod.txt)

**Comment by Andris-R** [comment]
Dear AMReX community,

I came across this discussion while investigating a similar issue. I have been using HDF5 (Chombo format) for several years, and I noticed that the current Fortran interface does not yet provide bindings for `amrex::WriteMultiLevelPlotfileHDF5` to write HDF5 plotfiles.

I was able to successfully build AMReX and run the `Advection_F` tutorial using the files uploaded by @LordOfBunnies. As a Fortran user, I would greatly appreciate seeing this functionality become part of the official Fortran interface in the future.

I understand that adding and maintaining new interface features introduces additional testing and maintenance overhead, but I wanted to express my support and interest in this feature.

@LordOfBunnies, thank you for sharing your implementation!

**Reply by WeiqunZhang** [reply]
See #5488.
